// @wdio/reporter/cjs/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.colorLines = exports.color = exports.pad = exports.getErrorsFromEvent = exports.sanitizeCaps = exports.sanitizeString = void 0;
const supportsColor_js_1 = __importDefault(require("./supportsColor.js"));
const constants_js_1 = require("./constants.js");
/**
 * replaces whitespaces with underscore and removes dots
 * @param  {string} str  variable to sanitize
 * @return {String}      sanitized variable
 */
function sanitizeString(str) {
    if (!str) {
        return '';
    }
    return String(str)
        .replace(/^.*\/([^/]+)\/?$/, '$1')
        .replace(/\./g, '_')
        .replace(/\s/g, '')
        .toLowerCase();
}
exports.sanitizeString = sanitizeString;
/**
 * formats capability object into sanitized string for e.g.filenames
 * @param {object} caps  Selenium capabilities
 */
function sanitizeCaps(caps) {
    if (!caps) {
        return '';
    }
    let result;
    /**
     * mobile caps
     */
    result = caps.deviceName
        ? [
            sanitizeString(caps.platformName),
            sanitizeString(caps.deviceName || caps['appium:deviceName']),
            sanitizeString(caps['appium:platformVersion']),
            sanitizeString(caps['appium:app'])
        ]
        : [
            sanitizeString(caps.browserName),
            sanitizeString(caps.version || caps.browserVersion),
            sanitizeString(caps.platform || caps.platformName),
            sanitizeString(caps['appium:app'])
        ];
    result = result.filter(n => n !== undefined && n !== '');
    return result.join('.');
}
exports.sanitizeCaps = sanitizeCaps;
/**
 * Takes a event emitted by a framework and extracts
 * an array of errors representing test or hook failures.
 * This exists to maintain compatibility between frameworks
 * with have a soft assertion model (Jasmine) and those that
 * have a hard assertion model (Mocha)
 * @param {*} e  An event emitted by a framework adapter
 */
function getErrorsFromEvent(e) {
    if (e.errors) {
        return e.errors;
    }
    if (e.error) {
        return [e.error];
    }
    return [];
}
exports.getErrorsFromEvent = getErrorsFromEvent;
/**
 * Pads the given `str` to `len`.
 *
 * @private
 * @param {string} str
 * @param {number} len
 * @return {string}
 */
function pad(str, len) {
    return Array(len - str.length + 1).join(' ') + str;
}
exports.pad = pad;
function color(type, content) {
    if (!supportsColor_js_1.default.stdout) {
        return String(content);
    }
    return `\u001b[${constants_js_1.COLORS[type]}m${content}\u001b[0m`;
}
exports.color = color;
/**
 * Colors lines for `str`, using the color `name`.
 *
 * @private
 * @param {string} name
 * @param {string} str
 * @return {string}
 */
function colorLines(name, str) {
    return str
        .split('\n')
        .map((str) => color(name, str))
        .join('\n');
}
exports.colorLines = colorLines;
