// @wdio/reporter/cjs/stats/suite.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const runnable_js_1 = __importDefault(require("./runnable.js"));
/**
 * Class describing statistics about a single suite.
 */
class SuiteStats extends runnable_js_1.default {
    uid;
    cid;
    file;
    title;
    fullTitle;
    tags;
    tests = [];
    hooks = [];
    suites = [];
    parent;
    /**
     * an array of hooks and tests stored in order as they happen
     */
    hooksAndTests = [];
    description;
    rule;
    constructor(suite) {
        super(suite.type || 'suite');
        this.uid = runnable_js_1.default.getIdentifier(suite);
        this.cid = suite.cid;
        this.file = suite.file;
        this.title = suite.title;
        this.fullTitle = suite.fullTitle;
        this.tags = suite.tags;
        this.parent = suite.parent;
        /**
         * only Cucumber
         */
        this.description = suite.description;
        this.rule = suite.rule;
    }
}
exports.default = SuiteStats;
