// @wdio/local-runner/build/repl.js
import WDIORepl from '@wdio/repl';
export default class WDIORunnerRepl extends WDIORepl {
    childProcess;
    callback;
    commandIsRunning = false;
    constructor(childProcess, options) {
        super(options);
        this.childProcess = childProcess;
    }
    _getError(params) {
        if (!params.error) {
            return null;
        }
        const err = new Error(params.message);
        err.stack = params.stack;
        return err;
    }
    eval(cmd, context, filename, callback) {
        if (this.commandIsRunning) {
            return;
        }
        this.commandIsRunning = true;
        this.childProcess.send({
            origin: 'debugger',
            name: 'eval',
            content: { cmd }
        });
        this.callback = callback;
    }
    onResult(params) {
        const error = this._getError(params);
        if (this.callback) {
            this.callback(error, params.result);
        }
        this.commandIsRunning = false;
    }
    start(context) {
        this.childProcess.send({
            origin: 'debugger',
            name: 'start'
        });
        return super.start(context);
    }
}
